### PR TITLE
[codegen/dotnet] Add support for ResourceType and isComponent

### LINKED ISSUE #5361: Generate component resource SDKs for .NET.
Depends on #5357.

### LINKED ISSUE #5413: [codegen/dotnet] Handle ResourceType
Depends on #5325

### PR DESCRIPTION
**Pull request by lblackstone**
Fix #5413
Fix #5361

Note that this PR depends on #5485

### CODE REVIEW

File: pkg/codegen/dotnet/gen.go [hunk 1]
```diff
@@ -242,6 +246,12 @@ func (mod *modContext) typeString(t schema.Type, qualifier string, input, state,
 		case mod.details(t).functionType:
 			typ += "Result"
 		}
+	case *schema.ResourceType:
+		typ = mod.tokenToResource(t.Token)
+		if typ != "" {
+			typ += "."
+		}
+		typ += tokenToName(t.Token)
```
Comment (justinvp) [inline pkg/codegen/dotnet/gen.go:254]: Any reason these extra lines aren't inside `tokenToResource`'s implementation? Or just not have `tokenToResource` and use `mod.tokenToNamespace(t.Token, "")` on line 250?

**Review by justinvp — APPROVED** [review]
LGTM

File: pkg/codegen/dotnet/gen.go [hunk 2]
```diff
@@ -628,13 +634,18 @@ func (mod *modContext) genResource(w io.Writer, r *schema.Resource) error {
 
 	// Open the class.
 	className := name
-	baseType := "Pulumi.CustomResource"
-	if mod.isK8sCompatMode() {
+	var baseType, optionsType string
+	switch {
+	case mod.isK8sCompatMode():
 		baseType = "KubernetesResource"
```
Comment (mikhailshilkov) [inline pkg/codegen/dotnet/gen.go:640]: Isn't `optionsType` needed for k8s?

File: pkg/codegen/dotnet/gen.go [hunk 3]
```diff
@@ -713,19 +722,28 @@ func (mod *modContext) genResource(w io.Writer, r *schema.Resource) error {
...
 	fmt.Fprintf(w, "        {\n")
 	fmt.Fprintf(w, "        }\n")
 
 	if mod.dictionaryConstructors {
-		fmt.Fprintf(w, "        internal %s(string name, ImmutableDictionary<string, object?> dictionary, CustomResourceOptions? options = null)\n", className)
-		fmt.Fprintf(w, "            : base(\"%s\", name, new DictionaryResourceArgs(dictionary), MakeResourceOptions(options, \"\"))\n", tok)
+		fmt.Fprintf(w, "        internal %s(string name, ImmutableDictionary<string, object?> dictionary, %s? options = null)\n",
+			className, optionsType)
+		if r.IsComponent {
+			fmt.Fprintf(w, "            : base(\"%s\", name, new DictionaryResourceArgs(dictionary), MakeResourceOptions(options, \"\"), remote: true)\n", tok)
```
Comment (mikhailshilkov) [inline pkg/codegen/dotnet/gen.go:737]: A dictionary constructor for a component doesn't make much sense to me, I would omit it for now.

**Review by mikhailshilkov — APPROVED** [review]
Looks great overall, two minor questions. It would be nice to see downstream changes before merging.

File: pkg/codegen/dotnet/gen.go (hunk 2 above)
Comment (lblackstone) [reply pkg/codegen/dotnet/gen.go:640]: Oops, good catch

**Comment by lblackstone** [comment]
The downstream AWS test is failing due to bad codegen changes related to imports. Working through those now.

**Comment by lblackstone** [comment]
Alright, downstream codegen shows no diffs so we should be good to go here! 🎉

**Pull request merged**